casper/src/test/scala/coop/rchain/casper/blocks/proposer/BlockReceiverSpec.scala

```diff
@@ -0,0 +1,161 @@
+package coop.rchain.casper.blocks.proposer
+
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import com.google.protobuf.ByteString
+import coop.rchain.casper.ValidatorIdentity
+import coop.rchain.casper.blocks.{BlockReceiver, BlockReceiverState}
+import coop.rchain.casper.helper.TestNode
+import coop.rchain.casper.helper.TestNode.Effect
+import coop.rchain.casper.protocol.{BlockMessage, Justification}
+import coop.rchain.casper.util.ConstructDeploy
+import coop.rchain.casper.util.GenesisBuilder.buildGenesis
+import coop.rchain.casper.util.scalatest.Fs2StreamMatchers
+import coop.rchain.crypto.signatures.Secp256k1
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.models.syntax._
+import coop.rchain.p2p.EffectsTestInstances.LogicalTime
+import coop.rchain.shared.Log
+import fs2.concurrent.Queue
+import monix.eval.Task
+import monix.testing.scalatest.MonixTaskTest
+import org.scalatest.Assertion
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+class BlockReceiverSpec
+    extends AsyncFlatSpec
+    with MonixTaskTest
+    with Matchers
+    with Fs2StreamMatchers {
+  implicit val logEff: Log[Task]            = Log.log[Task]
+  implicit val timeEff: LogicalTime[Effect] = new LogicalTime[Effect]
+  private val genesis                       = buildGenesis()
+
+  import fs2._
+
+  private def withBlockReceiverEnv(shardId: String)(
+      f: (
+          Queue[Task, BlockMessage],
+          Queue[Task, BlockMessage],
+          Stream[Task, BlockHash],
+          TestNode[Task]
+      ) => Task[Assertion]
+  ): Task[Assertion] = TestNode.standaloneEff(genesis).use { node =>
```

Comment:
This is way too much overhead to test the whole block receiver, not alone just its effects.
This means for BlockStore to be a mock with queues created on methods that we wan't collect and verify if called with correct arguments depending on input stream, or not called. E.g. KeyValueStore.put should be called when block is valid and stored by block receiver.

What just want to check if blocks from input stream are being supplied to dependencies like BlockStore and push to the output stream, or ignored if invalid.
The same is for BlockDagStorage and checking for existing blocks.

Of course, these tests don't need to repeat tests for internal state of the receiver.